test/addons-napi/test_promise/test_promise.c

```diff
@@ -0,0 +1,50 @@
+#include <stdio.h>
+#include <node_api.h>
+#include "../common.h"
+
+napi_value bind_napi_create_promise(napi_env env, napi_callback_info info) {
+  napi_value promise, resolve, reject, result;
+
+  NAPI_CALL(env, napi_create_promise(env, &promise, &resolve, &reject));
+  NAPI_CALL(env, napi_create_object(env, &result));
+
+  fprintf(stderr, "resolve: %p\n", resolve);
+
+  napi_valuetype value_type;
+  NAPI_CALL(env, napi_typeof(env, resolve, &value_type));
+  fprintf(stderr, "resolve: value_type: %d\n", value_type);
+
+  napi_property_descriptor properties[] = {
+    { "promise", NULL, NULL, NULL, NULL, promise, napi_default, NULL },
+    { "resolve", NULL, NULL, NULL, NULL, resolve, napi_default, NULL },
+    { "reject", NULL, NULL, NULL, NULL, reject, napi_default, NULL },
+  };
```

Comment:
I don't believe the primary use case is to return a promise to JS so that it may be resolved in JS. That can be accomplished without native code.

I believe the primary use case is to return a promise to JS that will be resolved from the native side - in which case you would never construct such an object. In fact, I only construct it here to make testing out the promise easier by allowing one to use JS to resolve/reject the promise.

OTOH, you're right in that you might have to pass the resolve and reject around the native code  such that they are always together, resulting in helper structures or a fatter stack. The deferred paradigm helps here too, because then you only have to pass one object around - the deferred.